plugin-server/src/worker/vm/upgrades/historical-export/export-historical-events-v2.ts

```diff
@@ -0,0 +1,487 @@
+- `exportHistoricalEvents` job is responsible for exporting data between particular start and end points
+    - It tracks its progress under `statusKey`
+    - It dynamically resizes the time window we fetch data to minimize waiting.
+    - It calls plugins `exportEvents` with each batch of events it finds
+    - It handles retries by retrying RetryErrors up to 15 times
+
+Note:
+- parallelism is only settable by superusers to avoid abuse.
+*/
+
+import { Plugin, PluginEvent, PluginMeta, RetryError } from '@posthog/plugin-scaffold'
+import * as Sentry from '@sentry/node'
+import { DateTime } from 'luxon'
+
+import {
+    Hub,
+    ISOTimestamp,
+    PluginConfig,
+    PluginConfigVMInternalResponse,
+    PluginLogEntrySource,
+    PluginLogEntryType,
+    PluginTask,
+    PluginTaskType,
+} from '../../../../types'
+import { isTestEnv } from '../../../../utils/env-utils'
+import { fetchEventsForInterval, fetchTimestampBoundariesForTeam, TimestampBoundaries } from '../utils/utils'
+
+const TEN_MINUTES = 1000 * 60 * 10
+const TWELVE_HOURS = 1000 * 60 * 60 * 12
+const EVENTS_TIME_INTERVAL = TEN_MINUTES
+export const EVENTS_PER_RUN = 500
+
+export const EXPORT_RUNNING_KEY = 'is_export_running_v2'
+
+const INTERFACE_JOB_NAME = 'Export historical events V2'
+
+export interface TestFunctions {
+    exportHistoricalEvents: (payload: ExportHistoricalEventsJobPayload) => Promise<void>
+    getTimestampBoundaries: (payload: ExportHistoricalEventsUIPayload) => Promise<TimestampBoundaries>
+    nextFetchTimeInterval: (payload: ExportHistoricalEventsJobPayload, eventCount: number) => number
+    coordinateHistoricExport: (update?: CoordinationUpdate) => Promise<void>
+    calculateCoordination: (
+        params: ExportParams,
+        done: Array<ISOTimestamp>,
+        running: Array<ISOTimestamp>
+    ) => Promise<CoordinationUpdate>
+    getExportDateRange: (params: ExportParams) => Array<[ISOTimestamp, ISOTimestamp]>
+    progressBar: (progress: number, length?: number) => string
+}
+
+export type ExportHistoricalEventsUpgradeV2 = Plugin<{
+    global: {
+        _testFunctions: TestFunctions
+    }
+}>
+
+export interface ExportHistoricalEventsJobPayload {
+    // Current cursor to what's being exported
+    timestampCursor: number
+
+    // The lower and upper bound of the timestamp interval to be processed
+    startTime: number
+    endTime: number
+
+    // The offset *within* a given timestamp interval
+    offset: number
+
+    // how many retries a payload has had (max = 15)
+    retriesPerformedSoFar: number
+
+    // used for ensuring only one "export task" is running if the server restarts
+    exportId: number
+
+    // Time frame to fetch events for.
+    fetchTimeInterval: number
+
+    // Key to report export status to
+    statusKey: string
+}
+
+export interface ExportHistoricalEventsUIPayload {
+    // Only set starting export from UI
+    parallelism?: number
```

Comment:
If we want only superusers to set this, is this the best place to have it?

I take it you'll add some validation on the Django side for this and then e.g. if we need to trigger an export for BIG_COMPANY we'd do it ourselves?

Don't have the answers, just wondering

Reply:
> I take it you'll add some validation on the Django side for this and then e.g. if we need to trigger an export for BIG_COMPANY we'd do it ourselves?

On cloud, exactly like this. Note that e.g. Simon and others can also be treated as staff for this purpose.